File: IOI '01 P1 - Mobile Phones.cpp
/*
https://dmoj.ca/problem/ioi01p1

2D Fenwick tree with basic query and update operations
- queries are 0-indexed, change to 1-indexing

*/

#include <bits/stdc++.h>
using namespace std;

class BIT_2D {
    int ROWS, COLS;
    vector<vector<int>> bit;

public:
    BIT_2D(int r, int c) : ROWS(r), COLS(c), bit(r + 1, vector<int>(c + 1)) {}

    void update(int r, int c, int diff) {
        for (; r <= ROWS; r += r & -r)
            for (int col = c; col <= COLS; col += col & -col)
                bit[r][col] += diff;
    }

    int query(int r, int c) {
        int total = 0;
        for (; r > 0; r -= r & -r)
            for (int col = c; col > 0; col -= col & -col)
                total += bit[r][col];
        return total;
    }

    int query_range(int r1, int c1, int r2, int c2) {
        return query(r2, c2) - query(r1 - 1, c2) - query(r2, c1 - 1) + query(r1 - 1, c1 - 1);
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int _, N;
    cin >> _ >> N;
    BIT_2D bit(N, N);

    while (true) {
        int q;
        cin >> q;
        if (q == 3)
            break;

        if (q == 1) {  // update operation
            int r, c, diff;
            cin >> r >> c >> diff;
            bit.update(r + 1, c + 1, diff);
        } else if (q == 2) {
            int r1, c1, r2, c2;
            cin >> r1 >> c1 >> r2 >> c2;
            cout << bit.query_range(r1 + 1, c1 + 1, r2 + 1, c2 + 1) << "\n";
        }
    }

    return 0;
}

File: IOI '01 P1 - Mobile Phones.py
"""
https://dmoj.ca/problem/ioi01p1
You can only submit this in c++ (check the other file)
python instantly MLEs

2D Fenwick tree with basic query and update operations
- queries are 0-indexed, change to 1-indexing

"""


class BIT_2D:  # uses 1-indexing
    def __init__(self, r, c):
        self.ROWS = r
        self.COLS = c
        self.bit = [[0] * (c + 1) for _ in range(r + 1)]

    def update(self, r, c, diff):
        while r <= self.ROWS:
            col = c
            while col <= self.COLS:
                self.bit[r][col] += diff
                col += col & (-col)
            r += r & (-r)

    def query(self, r, c):
        total = 0
        while r > 0:
            col = c
            while col > 0:
                total += self.bit[r][col]
                col -= col & (-col)
            r -= r & (-r)
        return total

    def query_range(self, r1, c1, r2, c2):
        return self.query(r2, c2) + \
            self.query(r1 - 1, c1 - 1) - \
            self.query(r1 - 1, c2) - \
            self.query(r2, c1 - 1)


_, N = map(int, input().split())
bit = BIT_2D(N, N)

while True:
    q = list(map(int, input().split()))
    if q[0] == 3:
        break

    if q[0] == 1:  # update operation
        r, c, diff = q[1:]
        bit.update(r + 1, c + 1, diff)

    elif q[0] == 2:
        r1, c1, r2, c2 = q[1:]
        print(bit.query_range(r1 + 1, c1 + 1, r2 + 1, c2 + 1))
